dcpy/connectors/edm/publishing.py

```diff
@@ -202,22 +204,53 @@ def publish(
     keep_draft: bool = True,
     max_files: int = s3.MAX_FILE_COUNT,
     latest: bool = False,
+    overwrite_if_exists: bool = False,
 ) -> None:
     """Publishes a specific draft build of a data product
     By default, keeps draft output folder"""
     if version is None:
         version = get_version(draft_key)
+
+    version_already_published = version in get_published_versions(
+        product=draft_key.product
+    )
+    # Bump version if input version already exists and overwrite isn't allowed
+    if version_already_published and not overwrite_if_exists:
+        version = versions.bump(
```

Comment:
I'm hesitant about this. I'd be confused if I called this function to publish a version, and then a different version was published. I think I'd just prefer it to throw an error letting me know that the version already exists.

Reply:
Hmm. Should `bump`ing a patch version then maybe _only_ happen when we're "repeating" a build? We have functionality around that, so maybe at time of `plan`, we have
- repeat from `build`/`draft` folder -> keep version of that build/draft
- repeat from `publish` folder -> bump `patch`

Reply:
@fvankrieken I'm intrigued at this idea of separating code vs input data in versions. So you could look at two versions side by side and see at a glance that the code was bumped, but not the input data (or vice versa). Not concrete thoughts - just intrigued.

Reply:
>  if a `patch` means code changes but not source data version changes

I think it could mean either. it seems like what we're going for with a patch is "oops, we have to re-distribute a release. the outside world will use the same version, but we don't want to overwrite our internal record of what was distributed the first time"

so whether because we had change code or we because we used the wrong version of source data, seems like either could be the motivation for a patch

Reply:
I think I agree that bumping for a patch should happen during the `plan` step. that's when we already bump the version via a product's `VersionStrategy` and the logic could go [here](https://github.com/NYCPlanning/data-engineering/blob/main/dcpy/lifecycle/builds/plan.py#L36)

seems like the plan logic could be something like:
```python
if this_is_a_patch:
  assert version_from_github_action is not None
  # bump the patch number of version_from_github_action and use that
else:
  if version_from_github_action is None:
    # use recipe.version_strategy to bump latest version and use that
  else:
    # use version_from_github_action
```

please forgive the nested `if/else`s, it's pseudocode. the logic above ignores `recipe.version` because we'd probably rather stop putting that in our checked-in recipe files

Reply:
just added a comment to the versions discussions [here](https://github.com/NYCPlanning/data-engineering/discussions/984#discussioncomment-10054943)

Reply:
I would like to revisit this discussion over a call/in-person to make sure I will be making correct changes to the PR

Reply:
So I think I went off track a little bit in my comments, and looking back I don't think significant change is needed.

Again, this is a rare case, and per our discussions in person, I think we intended to append a patch when needed, and while my idealized case sounds nice, I think @sf-dcp 's points (in chat outside of github) around how/when we might need to do this maybe means that something more like this will be easier/necessary.

I do think version should not be an option to supply here via gha - I think it only should be read with the logic in line 212. Manually specifying version at time of publishing seems like asking for manual errors.

Also to Alex's point - maybe it's a tad risky to just automatically "patch" if the version exists. To this point maybe its worth having a bool argument - `is_patch` or something like that. Without this specified, we throw error on version collision, while with it, we patch.

And I still think we should do away with overwriting as a possibility

Reply:
Add `is_patch` option: 

- if a version exists and `is_patch = True`, create a patched version
- if it doesn't and `is_patch = True or False`, create the normal version
-  if a version exists and `is_patch = False`, throw an error